aiida/calculations/transfer.py

```diff
@@ -0,0 +1,211 @@
+
+    The two main lists to include are `local_files` (for files to be taken from FolderData nodes)
+    and `remote_files` (for files to be taken from RemoteData nodes). Alternatively, files inside
+    of RemoteData nodes can instead be put in the `symlink_files` list: the only difference is that
+    files from the first list will be fully copied in the target RemoteData folder, whereas for the
+    files in second list only a symlink to the original file will be created there. This will only
+    affect the content of the final RemoteData target folder, but in both cases the full file will
+    be copied back in the local target FolderData (if `retrieve_files = True).
+    """
+
+    @classmethod
+    def define(cls, spec):
+        super().define(spec)
+
+        spec.input(
+            'instructions',
+            valid_type=orm.Dict,
+            help='A dictionary containing the `retrieve_files` flag and at least one of the file lists:'
+            '`local_files`, `remote_files` and/or `symlink_files`.',
+        )
+        spec.input_namespace(
+            'source_nodes',
+            valid_type=(orm.FolderData, orm.RemoteData),
+            dynamic=True,
+            help='All the nodes that contain files referenced in the instructions.',
+        )
+
+        # The transfer just needs a computer, the code are resources are set here
+        spec.inputs['code'].default = None
+        spec.inputs['code'].valid_type = (orm.Code, type(None))
+        spec.inputs['metadata']['computer'].required = True
+        spec.inputs['metadata']['options']['resources'].default = {
+            'num_machines': 1,
+            'num_mpiprocs_per_machine': 1,
+        }
+
+        # Retrieved node should only be generated if actually retrieving files
+        #spec.outputs[cls.link_label_retrieved].required = False
+        #spec.output(cls.link_label_retrieved).required = False
+
+        spec.inputs.validator = validate_transfer_inputs
+
+    def prepare_for_submission(self, folder):
+        source_nodes = self.inputs['source_nodes']
+        instructions = self.inputs['instructions'].get_dict()
+        retrieve_files = instructions.get('retrieve_files')
+
+        calcinfo = CalcInfo()
+        calcinfo.codes_info = []
+        calcinfo.retrieve_list = []
+
+        calcinfo.local_copy_list = []
+        local_files = instructions.get('local_files', [])
+        for source_label, source_relpath, target_relpath in local_files:
+
+            source_node = source_nodes[source_label]
+            if not isinstance(source_node, orm.FolderData):
+                raise RuntimeError(
+                    'type of node {} for file {} in `local_files` list'
+                    'must be orm.FolderData instead of {}'.format(
+                        source_label,
+                        source_relpath,
+                        source_node.node_type,
+                    )
+                )
```

Comment:
I did think about this, but I thought that since here I'm cycling through the files here and not in the validation it was better this way than doing double the cycles. I guess that this could be a premature optimization, and also I could maybe explicitly do the conversions of lists into sets and do this check there instead of using the list compression and technically keep the same number of cycles...

Reply:
Good point, but in this case this performance hit is acceptable, because know, if the value is incorrect, you will except in `prepare_for_submission`. At that point the node will already have been stored in the database. I think it is preferable to have slight inefficiency in order to prevent unnecessary excepted nodes in the database